tensorflow_addons/utils/test_utils.py

```diff
@@ -53,6 +53,75 @@ def use_gpu():
         yield
 
 
+def create_virtual_devices(num_devices,
+                           force_device=None,
+                           memory_limit_per_device=1024):
+    """Virtualize a the physical device into logical devices.
+
+    Args:
+        num_devices: The number of virtual devices needed.
+        force_device: 'CPU'/'GPU'. Defaults to None, where the
+            devices is selected based on the system.
+        memory_limit_per_device: Specify memory for each
+            virtual GPU. Only for GPUs.
+
+    Returns:
+        virtual_devices: A list of virtual devices which can be passed to
+            tf.distribute.MirroredStrategy()
+    """
+    if force_device is None:
+        device_type = 'GPU' if len(
+            tf.config.list_physical_devices('GPU')) > 0 else 'CPU'
+    else:
+        assert (force_device in ['CPU', 'GPU'])
+        device_type = force_device
+
+    physical_devices = tf.config.list_physical_devices(device_type)
+
+    if device_type == 'CPU':
+        memory_limit_per_device = None
+
+    tf.config.experimental.set_virtual_device_configuration(
+        physical_devices[0], [
+            tf.config.experimental.VirtualDeviceConfiguration(
+                memory_limit=memory_limit_per_device)
+            for _ in range(num_devices)
+        ])
+
+    return tf.config.experimental.list_logical_devices(device_type)
+
+
+def run_all_distributed(num_devices):
+    base_decorator = run_distributed(num_devices)
+
+    def decorator(cls):
+        for name, method in cls.__dict__.copy().items():
+            if (callable(method)
+                    and name.startswith(unittest.TestLoader.testMethodPrefix)
+                    and name != "test_session"):
+                setattr(cls, name, base_decorator(method))
+        return cls
+
+    return decorator
+
+
+def run_distributed(num_devices):
+    def decorator(f):
+        if inspect.isclass(f):
+            raise TypeError("`run_distributed` only supports test methods. "
+                            "Did you mean to use `run_all_distributed`?")
+
+        def decorated(self, *args, **kwargs):
+            logical_devices = create_virtual_devices(num_devices)
+            strategy = tf.distribute.MirroredStrategy(logical_devices)
```

Comment:
Is there a way we can make the distribution strategy configurable? This might be useful for certain types of bugs

Reply:
Nope, each strategy uses different constructor params. I can have them pass a class to the decorated and instantiate that with `logical_devices` and throw if there's an exception.

Reply:
I'll leave the support for other strategies as a TODO, and have this in the repo for now. `MirroredStrategy` is the most widely used in the community as well.